### PR TITLE
Unsafe ops for QQ/ZZPolyRingElem, QQ/ZZMatrix

### PR DESCRIPTION
**Pull request by fingolfin**
Move flintify calls down into unsafe ops

### CODE REVIEW

**Comment by fingolfin** [comment]
Ah it seems this may require PR #1905. Let's try again once that one is merged.

File: src/flint/fmpq_poly.jl [hunk 1]
```diff
@@ -772,54 +652,178 @@ end
...
+end
+
+function sub!(z::QQPolyRingElemOrPtr, x::Int, y::QQPolyRingElemOrPtr)
+  @ccall libflint.fmpq_poly_si_sub(z::Ref{QQPolyRingElem}, x::Int, y::Ref{QQPolyRingElem})::Nothing
+  return z
+end
+
+sub!(z::QQPolyRingElemOrPtr, x::QQPolyRingElemOrPtr, y::Union{Integer, Rational}) = sub!(z, x, flintify(y))
+
+sub!(z::QQPolyRingElemOrPtr, x::Union{Integer, Rational}, y::QQPolyRingElemOrPtr) = sub!(z, flintify(x), y)
```
Comment (lgoettgens) [inline src/flint/fmpq_poly.jl:799]: same problem with `UInt`

File: src/flint/fmpq_mat.jl [hunk 2]
```diff
@@ -302,90 +295,39 @@ end
...
+    function +(mat::QQMatrix, scalar::$T)
+      z = deepcopy(mat)
+      for i = 1:min(nrows(mat), ncols(mat))
+        add!(mat_entry_ptr(z, i, i), scalar)
       end
       return z
     end
 
-    +(x::$T, y::QQMatrix) = y + x
+    +(mat::$T, scalar::QQMatrix) = scalar + mat
```
Comment (lgoettgens) [inline src/flint/fmpq_mat.jl:311]:
```suggestion
    +(scalar::$T, mat::QQMatrix) = mat + scalar
```

File: src/flint/fmpz_mat.jl [hunk 3]
```diff
@@ -312,66 +312,37 @@ end
...
+    function +(mat::ZZMatrix, scalar::$T)
+      z = deepcopy(mat)
+      for i = 1:min(nrows(mat), ncols(mat))
+        add!(mat_entry_ptr(z, i, i), scalar)
+      end
+      return z
+    end
 
--(x::ZZMatrix, y::ZZRingElem) = x + (-y)
+    +(mat::$T, scalar::ZZMatrix) = scalar + mat
```
Comment (lgoettgens) [inline src/flint/fmpz_mat.jl:328]:
```suggestion
    +(scalar::$T, mat::ZZMatrix) = mat + scalar
```

File: src/flint/fmpz_mat.jl [hunk 4]
```diff
@@ -1810,81 +1781,91 @@ end
...
+function mul!(z::ZZMatrixOrPtr, a::ZZMatrixOrPtr, b::ZZRingElemOrPtr)
   ccall((:fmpz_mat_scalar_mul_fmpz, libflint), Nothing,
-        (Ref{ZZMatrix}, Ref{ZZMatrix}, Ref{ZZRingElem}), z, y, x)
+        (Ref{ZZMatrix}, Ref{ZZMatrix}, Ref{ZZRingElem}), z, a, b)
   return z
 end
 
-function addmul!(z::ZZMatrix, y::ZZMatrix, x::ZZRingElem)
+mul!(z::ZZMatrixOrPtr, a::ZZMatrixOrPtr, b::Integer) = mul!(z, a, flintify(b))
+mul!(z::ZZMatrixOrPtr, a::Union{Integer, ZZRingElemOrPtr}, b::ZZMatrixOrPtr) = mul!(z, b, a)
```
Comment (lgoettgens) [inline src/flint/fmpz_mat.jl:1838]:
```suggestion
mul!(z::ZZMatrixOrPtr, a::Union{IntegerUnionOrPtr}, b::ZZMatrixOrPtr) = mul!(z, b, a)
```

File: src/flint/fmpz_mat.jl [hunk 5]
```diff
@@ -1810,81 +1781,91 @@ end
...
 
-function mul!(z::Vector{ZZRingElem}, a::ZZMatrix, b::Vector{ZZRingElem})
+function addmul!(z::ZZMatrixOrPtr, a::ZZMatrixOrPtr, b::UInt)
+  ccall((:fmpz_mat_scalar_addmul_ui, libflint), Nothing,
+        (Ref{ZZMatrix}, Ref{ZZMatrix}, UInt), z, a, b)
+  return z
+end
+
+addmul!(z::ZZMatrixOrPtr, a::ZZMatrixOrPtr, b::Integer) = addmul!(z, a, flintify(b))
+addmul!(z::ZZMatrixOrPtr, a::Union{Integer, ZZRingElemOrPtr}, b::ZZMatrixOrPtr) = addmul!(z, b, a)
```
Comment (lgoettgens) [inline src/flint/fmpz_mat.jl:1859]:
```suggestion
addmul!(z::ZZMatrixOrPtr, a::Union{IntegerUnionOrPtr}, b::ZZMatrixOrPtr) = addmul!(z, b, a)
```

File: src/flint/fmpz_mat.jl (hunk 4 above)
Comment (fingolfin) [reply src/flint/fmpz_mat.jl:1838]: You probably meant this?
```suggestion
mul!(z::ZZMatrixOrPtr, a::IntegerUnionOrPtr, b::ZZMatrixOrPtr) = mul!(z, b, a)
```

File: src/flint/fmpz_mat.jl (hunk 5 above)
Comment (fingolfin) [reply src/flint/fmpz_mat.jl:1859]:
```suggestion
addmul!(z::ZZMatrixOrPtr, a::IntegerUnionOrPtr, b::ZZMatrixOrPtr) = addmul!(z, b, a)
```

File: src/flint/fmpq_mat.jl (hunk 2 above)
Comment (fingolfin) [reply src/flint/fmpq_mat.jl]: Ouch, thanks, well-spotted

File: src/flint/fmpz_mat.jl (hunk 4 above)
Comment (lgoettgens) [reply src/flint/fmpz_mat.jl]: oops, yeah of course

**Pull request merged**